File: Ruby/CivitasR/lib/Jugador.rb
#encoding:utf-8

module Civitas
  require_relative 'Diario'
  require_relative 'Dado'

  class Jugador

    include Comparable

    @@CasasMax = 4
    @@CasasPorHotel = 4
    @@HotelesMax = 4
    @@PasoPorSalida = 1000.0
    @@PrecioLibertad = 200.0
    @@SaldoInicial = 7500.0


    def initialize(name)
      @nombre=name
      @encarcelado = false
      @salvoconducto = nil
      @puedeComprar = true
      @propiedades = []
      @numCasillaActual = 0
      @saldo = @@SaldoInicial
    end

    def self.copia(tjugador)
      jugador = new(tjugador.nombre)
      jugador.encarcelado = tjugador.encarcelado
      jugador.salvoconducto = tjugador.salvoconducto
      jugador.puedeComprar = tjugador.puedeComprar
      jugador.propiedades = tjugador.propiedades
      jugador.numCasillaActual = tjugador.numCasillaActual
      return jugador
    end

    def cancelarHipoteca(ip)
      result=false
      if @encarcelado
        return result
      end
      if existeLaPropiedad(ip)
        propiedad=@propiedades.get(ip);
        cantidad=propiedad.getImporteCancelarHipoteca
        puedoGastar=puedoGastar(cantidad)
        if puedoGastar
          result=propiedad.cancelarHipoteca(self)
          if result
            Diario.ocurreEvento("El jugador "+@nombre+" cancela la hipoteca de la propiedad "+ip.to_s)
          end
        end
      end
      return result
    end
    private :cancelarHipoteca

    def puedoEdificarHotel(propiedad)
      return puedoGastar(propiedad.getPrecioEdificar) && (propiedad.numHoteles<@@HotelesMax) && (propiedad.numCasas>=@@CasasPorHotel)
    end
    
    def construirHotel(ip)
      result=false
      if @encarcelado
        return result
      end
      if existeLaPropiedad(ip)
        propiedad = @propiedades.at(ip)
        if puedoEdificarHotel(propiedad)
          result=propiedad.construirHotel(self)
          casasPorHotel=@@CasasPorHotel
          propiedad.derruirCasas(casasPorHotel,self)
        end
        Diario.instance.ocurre_evento("El jugador "+ toString + " construye hotel en la propiedad "+ ip.toString)
      end
      return result
    end
    
    def puedoEdificarCasa(propiedad)
      return puedoGastar(propiedad.precioEdificar) && (propiedad.numCasas < @@CasasMax)
    end
    
    def construirCasa(ip)
      result = false
      if @encarcelado
        return result
      end
      if existeLaPropiedad(ip)
        propiedad = @propiedades[ip]
        if puedoEdificarCasa(propiedad)
          result = propiedad.construirCasa(self)
          if result
            Diario.instance.ocurre_evento(
              "El jugador " + toString + " construye casa en la propiedad "+ propiedad.toString
            )
          end
        end
      end
      return result
    end

    def debeSerEncarcelado
      if(@encarcelado)
        resul=false
      else
        if(tieneSalvoconducto)
          perderSalvoconducto
          Diario.instance.ocurre_evento("Jugador -- El jugador " + toString + " se libra de ir a la carcel.")
          resul=false
        else
          resul=true
        end
      end
      return resul 
    end
    protected :debeSerEncarcelado

    def encarcelar(numCasillaCarcel)
      if(debeSerEncarcelado)
        moverACasilla(numCasillaCarcel)
        @encarcelado=true
        Diario.instance.ocurre_evento("Jugador -- El jugador " + toString + " ha sido encarcelado")
      end
      return @encarcelado
    end

    def obtenerSalvoconducto(s)
      if(!@encarcelado)
        @salvoconducto=s
        resul=true
      else
        resul=false
      end
      return resul
    end 

    def perderSalvoconducto
      @salvoconducto.usada
      @salvoconducto=nil
    end
    private :perderSalvoconducto

    def tieneSalvoconducto 
      @salvoconducto != nil
    end

    def puedeComprarCasilla
      @puedeComprar = !@encarcelado
      return @puedeComprar
    end

    def paga(cantidad)
      return modificarSaldo(-1*cantidad)
    end

    def pagaImpuesto(cantidad)
      @encarcelado ? false : paga(cantidad)
    end

    def pagaAlquiler(cantidad)
      @encarcelado ? false : paga(cantidad)
    end

    def recibe(cantidad)
      @encarcelado ? false : modificarSaldo(cantidad)
    end

    def modificarSaldo(cantidad)
      @saldo+=cantidad
      Diario.instance.ocurre_evento("Jugador -- Se ha modificado el saldo de " + toString + " en " + cantidad.to_s + " euros.")
      return true
    end

    def moverACasilla(numCasilla)
      if(@encarcelado)
        return false
      else
        @numCasillaActual=numCasilla
        @puedeComprar=false
        Diario.instance.ocurre_evento("Jugador -- Se ha movido al jugador a la casilla "+@numCasillaActual.to_s)
        return true
      end
    end

    def puedoGastar(precio)
      @encarcelado ? false : @saldo>=precio
    end
    private :puedoGastar

    def vender(ip)
      if(@encarcelado)
        return false
      end
      if(existeLaPropiedad(ip))
        resul=@propiedades.at(ip).vender(self)
        if resul
          @propiedades.delete_at(ip)
          Diario.ocurre_evento("Vendida la propiedad : "+ip.to_s)
          return true
        end
      end
      return false
    end

    def comprar(titulo)
      result = false
      if(@encarcelado)
        return result
      end
      if @puedeComprar
        precio = titulo.precioCompra
        if puedoGastar(precio)
          result = titulo.comprar(self)
          if result
            @propiedades << titulo
            Diario.instance.ocurre_evento("El jugador " + @nombre + " compra la propiedad " + titulo.toString);
          end
          @puedeComprar = false
        end
      end
      return result
    end

    def hipotecar(ip)
      result = false
      if @encarcelado
        return result
      end
      if existeLaPropiedad(ip)
        result = @propiedades[ip].hipotecar(self)
      end
      if result
        Diario.instance.ocurre_evento("El jugador " + @jugador.getNombre + " compra la propiedad " + titulo.toString)  
      end
    end

    def tieneAlgoQueGestionar
      return getPropiedades.empty?
    end

    def puedeSalirCarcelPagando
      return @saldo>=@@PrecioLibertad
    end
    private :puedeSalirCarcelPagando

    def salirCarcelPagando
      if(@encarcelado && puedeSalirCarcelPagando)
        paga(@@PrecioLibertad)
        @encarcelado=false
        Diario.instance.ocurre_evento("Jugador -- El jugador " + toString + " paga para salir de la carcel.")
        return true
      else
        return false
      end
    end

    def salirCarcelTirando
      if(Dado.salgoDeLaCarcel)
        @encarcelado=false
        Diario.instance.ocurre_evento("El jugador " + toString + " sale de la carcel tirando dados.")
        return true
      else
        return false
      end
    end

    def pasaPorSalida
      modificarSaldo(@@PasoPorSalida)
      Diario.ocurre_evento("El jugador ha pasado por salida.")
      return true
    end 

    def <=> (otro)
      @saldo <=> otro.saldo
    end



    #   Metodos que no estan en el guion pero si
    #   en el diagrama estructural.



    def enBancarrota
      @saldo<0
    end

    def existeLaPropiedad(ip)
      @propiedades.size > ip
    end
    private :existeLaPropiedad


    def puedeEdificarCasa(propiedad)
      return propiedad.getNumCasas<@@CasasMax
    end
    private :puedeEdificarCasa

    def puedeEdificarHotel(propiedad)
      return (propiedad.numCasas==4 && propiedad.numHoteles < @@HotelesMax)
    end
    private :puedeEdificarHotel

    def toString
      return @nombre + " (Saldo: " + @saldo.to_s + "; Casilla: " + @numCasillaActual.to_s + 
             "; Propiedades: " + @propiedades.size().to_s + "; Encarcelado: " + @encarcelado.to_s + ")"
    end
    
    def cantidadCasasHoteles
      i = 0
      for p in @propiedades
        i += p.numCasas + p.numHoteles
      end
      return i
    end

    attr_reader :CasasPorHotel,:encarcelado,:puedeComprar,:numCasillaActual

    private 
      attr_reader  :HotelesMax,:CasasMax,:PasoPorSalida,:PrecioLibertad

    protected
      attr_reader :nombre,:saldo,:propiedades


  end
end

File: Java/Civitas/src/civitas/Jugador.java
package civitas;
import java.lang.Math;
import GUI.Dado;
import java.util.ArrayList;

public class Jugador implements Comparable<Jugador>{
    protected static int CasasMax = 4;
    protected static int CasasPorHotel = 4;
    protected static int HotelesMax = 4;
    protected static float PasoPorSalida = 1000;
    protected static float PrecioLibertad = 200;
    protected static float SaldoInicial = 7500;
    
    protected boolean encarcelado;
    protected String nombre;
    protected int numCasillaActual;
    protected boolean puedeComprar;
    protected float saldo;
    protected Sorpresa_SalirCarcel salvoconducto;
    protected ArrayList<TituloPropiedad> propiedades;

    public Jugador(String tnombre){
            nombre = tnombre;
            encarcelado = false;
            numCasillaActual = 0;
            puedeComprar = true;
            saldo = SaldoInicial;
            salvoconducto = null;
            propiedades = new ArrayList<>();
    }

    public Jugador(Jugador otro){
            nombre = otro.nombre;
            encarcelado = otro.encarcelado;
            numCasillaActual = otro.numCasillaActual;
            puedeComprar = otro.puedeComprar;
            saldo = otro.saldo;
            salvoconducto = otro.salvoconducto;
            propiedades = new ArrayList<>(otro.propiedades);		
    }

    public Boolean isEncarcelado(){
            return encarcelado;
    }

    protected int getCasasMax(){
            return CasasMax;
    }
    int getCasasPorHotel(){
            return CasasPorHotel;
    }
    protected int getHotelesMax(){
            return HotelesMax;
    }
    public String getNombre(){
            return nombre;
    }
    int getNumCasillaActual(){
            return numCasillaActual;
    }
    private float getPrecioLibertad(){
            return PrecioLibertad;
    }
    private float getPremioPasoSalida(){
            return PasoPorSalida;
    }

    public ArrayList<TituloPropiedad> getPropiedades(){
            return propiedades;
    }
    boolean getPuedeComprar(){
            return puedeComprar;
    }
    public float getSaldo(){
            return saldo;
    }
    boolean enBancarrota(){
            return saldo < 0;
    }

    int cantidadCasasHoteles(){
            int sum = 0;
            for(TituloPropiedad i:propiedades){
                    sum += i.getNumCasas()+i.getNumHoteles();
            }
            return sum;
    }

    boolean comprar(TituloPropiedad titulo){
            boolean result=false;
            if(encarcelado){
                    return result;
            }
            if(puedeComprar){
                    float precio=titulo.getPrecioCompra();
                    if(puedoGastar(precio)){
                            result=titulo.comprar(this);
                            if(result){
                                    propiedades.add(titulo);
                                    Diario.getInstance().ocurreEvento("El jugador "+nombre+" compra la propiedad "+ titulo.toString());
                            }
                            puedeComprar=false;
                    }
            }
            return result;
    }
	
        
	
    protected boolean debeSerEncarcelado(){
        if(!encarcelado && tieneSalvoconducto()){
            Diario.getInstance().ocurreEvento("Jugador -- "
                + nombre + " usa su carta sorpresa para evitar la carcel");
            perderSalvoconducto();
            return false;
        }
        return !encarcelado;
    }
    
    boolean encarcelar(int numCasillaCarcel){
        if(debeSerEncarcelado()){
            moverACasilla(numCasillaCarcel);
            encarcelado = true;
            Diario.getInstance().ocurreEvento("Jugador -- "
                    + nombre + " es mandado a la carcel");
        }
        return encarcelado;
    }
	
    boolean obtenerSalvoconducto(Sorpresa_SalirCarcel s){
            if(!encarcelado){
                    salvoconducto = s;
            }
            return !encarcelado;
    }

    protected void perderSalvoconducto(){
            salvoconducto.usada();
            salvoconducto = null;
    }

    boolean tieneSalvoconducto(){
            return salvoconducto != null;
    }

    boolean puedeComprarCasilla(){
            puedeComprar=(!encarcelado);
            return puedeComprar;
    }

    boolean modificarSaldo(float cantidad){
            saldo += cantidad;
            return true;
    }

    boolean paga(float cantidad){
            return modificarSaldo(cantidad * -1);
    }

    boolean pagaImpuesto(float cantidad){
            return !encarcelado && paga(cantidad);
    }

    boolean pagaAlquiler(float cantidad){
            return !encarcelado && paga(cantidad);
    }

    boolean recibe(float cantidad){
            return !encarcelado && modificarSaldo(cantidad);
    }

    boolean moverACasilla(int numCasilla){
            if(!encarcelado){
                    numCasillaActual = numCasilla;
                    puedeComprar = false;
                    Diario.getInstance().ocurreEvento("Jugador -- "
                                    + nombre + " se ha movido a la casilla " + String.valueOf(numCasilla));
                    return true;
            }
            else{
                return encarcelado;
            }
    }

    private boolean puedoGastar(float precio){
            return !encarcelado && saldo >= precio;
    } 

    private boolean existeLaPropiedad(int ip){
            return ip < propiedades.size();
    }

    boolean vender(int ip){
            if(encarcelado || !existeLaPropiedad(ip) || !propiedades.get(ip).vender(this)){
                    return false;
            }
            Diario.getInstance().ocurreEvento("Jugador -- "
                            + nombre + " ha vendido la casilla " + propiedades.get(ip).getNombre());
            propiedades.remove(ip);
            return true;
    }


    boolean tieneAlgoQueGestionar(){
            return !propiedades.isEmpty();
    }

    boolean salirCarcelPagando(){
            if(encarcelado && paga(PrecioLibertad)){
                    encarcelado = false;
                    Diario.getInstance().ocurreEvento("Jugador -- "
                                    + nombre + " paga para salir de la carcel");
                    return true;
            } else {
                    return false;
            }
    }

    boolean salirCarcelTirando(){
            boolean b = Dado.getInstance().salgoDeLaCarcel();
            if(b){
                    encarcelado = false;
                    Diario.getInstance().ocurreEvento("Jugador -- "
                                    + nombre + " consigue salir de la carcel tirando dados");
            }
            return b;
    }


    boolean pasaPorSalida(){
            saldo += PasoPorSalida;
            Diario.getInstance().ocurreEvento("Jugador -- "
                            + nombre + " ha pasado por la casilla de salida");
            return true;
    }

    public String toString(){
            String resul="Jugador: " + nombre + "; Posicion: " + String.valueOf(numCasillaActual) + "; Saldo: " + String.valueOf(getSaldo())+
                         "; Propiedades: " + String.valueOf(propiedades.size()) + "; Encarcelado: " + Respuestas.values()[encarcelado ? 1:0] +"\n";
            return resul;
    }

    public String toStringPropiedades(){
            String resul="Propiedades:\n";
            for(int i=0;i<propiedades.size();i++){
                    resul+=propiedades.get(i).toString() +"\n";
            }
            return resul+"\n";
    }

    public int compareTo(Jugador otro){
            return (int) Math.signum(saldo - otro.saldo);
    }

    boolean cancelarHipoteca(int ip){
            boolean result = false;
            if( encarcelado ){
                    return result;
            }
            if( existeLaPropiedad(ip) ){
                    TituloPropiedad propiedad = propiedades.get(ip);
                    if( puedoGastar( propiedad.getImporteCancelarHipoteca() ) ){
                            result = propiedad.cancelarHipoteca(this);
                            if( result ){
                                    Diario.getInstance().ocurreEvento("Jugador -- El jugador " + 
                                                    toString() + " cancela la hipoteca de la propiedad " + propiedad.toString());
                            }
                    }
            }
            return result;
    }

    private boolean puedoEdificarCasa(TituloPropiedad prop){
            return  puedoGastar( prop.getPrecioEdificar() ) &&
                            prop.getNumCasas() < this.getCasasMax();
    }

    boolean construirCasa(int ip){
            boolean result=false;
            if(!encarcelado){	//1
                    boolean existe=existeLaPropiedad(ip); //2
                    if(existe){
                            TituloPropiedad propiedad=propiedades.get(ip); //3
                            if(this.puedoEdificarCasa(propiedad)){
                                    result=propiedad.construirCasa(this);//5
                                    if(result)
                                            Diario.getInstance().ocurreEvento("\nEl jugador -- " + toString()  + 
                                                    " construye casa en la propiedad "+ propiedad.toString()); //6
                            }
                    }
            }
            return result; //1
    }

    private boolean puedoEdificarHotel(TituloPropiedad prop){
            return  puedoGastar( prop.getPrecioEdificar() ) &&
                            prop.getNumHoteles() < this.getHotelesMax() &&
                            prop.getNumCasas() >= CasasPorHotel;
    }

    boolean construirHotel(int ip){
            boolean result = false;
            if(encarcelado){
                    return result;
            }
            if( existeLaPropiedad(ip) ){
                    TituloPropiedad prop = propiedades.get(ip);
                    if(puedoEdificarHotel(prop)){
                            result = prop.construirHotel(this);
                            prop.derruirCasas( CasasPorHotel , this);
                            if(result) 
                                Diario.getInstance().ocurreEvento("\nJugador --  "  + toString()  + 
                                            " construye hotel en la propiedad " + prop.toString());
                    }
            }
            return result;
    }

    boolean hipotecar(int ip) {
       	boolean result=false;
		TituloPropiedad propiedad;
	   	if(encarcelado){
			return result;
	   	}
		if(existeLaPropiedad(ip)){
			propiedad=propiedades.get(ip);
			result=propiedad.hipotecar(this);
			if(result){
				Diario.getInstance().ocurreEvento("El jugador "+ toString() +
							" hipoteca la propiedad "+ propiedad.toString() );
			}
		}
		return result;
	}

}